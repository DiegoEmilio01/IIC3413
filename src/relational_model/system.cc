#include "system.h"

#include <stdexcept>

// memory for the global objects
static typename std::aligned_storage<sizeof(BufferManager), alignof(BufferManager)>::type buffer_mgr_buf;
static typename std::aligned_storage<sizeof(FileManager), alignof(FileManager)>::type     file_mgr_buf;
static typename std::aligned_storage<sizeof(Catalog), alignof(Catalog)>::type             catalog_buf;

BufferManager& buffer_mgr  = reinterpret_cast<BufferManager&>(buffer_mgr_buf);
FileManager& file_mgr      = reinterpret_cast<FileManager&>(file_mgr_buf);
Catalog& catalog           = reinterpret_cast<Catalog&>(catalog_buf);


System::System(const std::string& db_folder, uint64_t buffer_size) {
    new (&file_mgr) FileManager(db_folder);
    new (&buffer_mgr) BufferManager(buffer_size);
    new (&catalog) Catalog("catalog.dat");
}

System System::init(const std::string& db_folder, uint64_t buffer_size) {
    if (initialized) {
        throw std::logic_error("System::init cannot be called twice");
    }
    initialized = true;
    return System(db_folder, buffer_size);
}

System::~System() {
    catalog.~Catalog();
    buffer_mgr.~BufferManager();
    file_mgr.~FileManager();
}