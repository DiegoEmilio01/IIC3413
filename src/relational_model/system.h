#pragma once

#include "relational_model/catalog.h"
#include "storage/buffer_manager.h"
#include "storage/file_manager.h"

// class SystemScope {
// public:
//     ~SystemScope();
// };

class System {
public:
    ~System();

    // static SystemScope init();
    static System init(const std::string& db_folder, uint64_t buffer_size);

private:
    static inline bool initialized = false;

    System(const std::string& db_folder, uint64_t buffer_size);
};


// global objects
extern FileManager& file_mgr;
extern BufferManager& buffer_mgr;
extern Catalog& catalog;
