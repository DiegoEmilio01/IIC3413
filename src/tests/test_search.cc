#include <iostream>

#include "relational_model/system.h"

constexpr uint64_t GB = 1024 * 1024 * 1024; // 1 GB

int main() {
    uint64_t buffer_size = 1 * GB;
    std::string database_folder = "data/test_example";

    // Need to call System::init before start using the database
    // When this object comes out of scope the database is no longer usable
    auto system = System::init(database_folder, buffer_size);

    Schema table1_schema(
        { "col1", "col2", "col3" },
        { DataType::INT, DataType::STR, DataType::INT }
    );

    Schema existing_table_schema;

    std::string table1 = "test_3";
    HeapFile* table = catalog.get_table(table1, &existing_table_schema);

    if (table == nullptr) { // table doesn't exist
        table = catalog.create_table(table1, table1_schema);
        catalog.insert_record(table1, {1, "record_1", 1});
        catalog.insert_record(table1, {2, "record_2", 2});
        catalog.insert_record(table1, {3, "record_21", 3});

        catalog.create_non_clustered_isam(table1, 1);

    } else {
        assert(existing_table_schema == table1_schema);
    }

    auto isam = catalog.get_index(table1);
    assert(isam != nullptr);

    auto iter = isam->get_iter(Value("record_1"), Value("record_2")); // test Value (not key) comparison

    Record& record_buf = catalog.get_record_buf(table1);
    iter->begin(record_buf);

    while (iter->next()) {
        // print result
        std::cout << record_buf << "\n";
    }

    return 0;
}
