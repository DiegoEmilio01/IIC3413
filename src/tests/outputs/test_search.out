1,"record_1",1
2,"record_2",2
